impactlab_tools/utils/configdict.py

```diff
@@ -0,0 +1,163 @@
+"""Class for representing tool configuration files
+"""
+
+import collections
+import inspect
+
+
+def gather_configtree(d):
+    """Chains nested-dicts into a connected tree of ConfigDict(s)
+
+    Parameters
+    ----------
+    d : dict
+        Cast to :py:class:`ConfigDict`. Nested dicts within are also
+        recursively cast and assigned parents, reflecting their nested
+        structure.
+
+    Returns
+    -------
+    out : ConfigDict
+
+    Examples
+    --------
+    .. code-block:: python
+
+        >>> nest = {'a': 1, 'b': {'a': 2}, 'c': 3, 'd-4': 4, 'e_5': 5, 'F': 6}
+        >>> tree = gather_configtree(nest)
+        >>> tree['b']['d-4']
+        4
+    """
+    out = ConfigDict(d)
+    for k, v in out.data.items():
+        # Replace nested maps with new ConfigDicts
+        if isinstance(v, collections.abc.MutableMapping):
+            out.data[k] = gather_configtree(v)
+            out.data[k].parent = out
+    return out
+
+
+class ConfigDict(collections.UserDict):
+    """Chain-able dictionary to hold projection configurations.
+
+    A ConfigDict is a dictionary-like interface to a chainmap/linked list.
+    Nested dicts can be access like a traditional dictionary but it searches
+    parent dictionaries for keys:values not found. All string keys normalized,
+    by transforming all characters to lowercase, and all underscores to
+    hyphens.
+
+    Attributes
+    ----------
+    parent : ConfigDict or None
+        Parent ConfigDict object to query for keys if not in `self.data`.
+    key_access_stack : dict
+        Dictionary with values giving the :py:func:`inspect.stack()` from the
+        most recent time a key was retrieved (via `self.__getitem__()`).
+    data : dict
+        The 'local' dictionary, not in parents.
+
+    See Also
+    --------
+    gather_configtree : Chains nested-dicts into a connected tree of
+        ConfigDict(s)
+
+    Examples
+    --------
+    .. code-block:: python
+
+        >>> d = {'a': 1, 'b': {'a': 2}, 'c': 3, 'd-4': 4, 'e_5': 5, 'F': 6}
+        >>> cd = ConfigDict(d)
+        >>> cd['b']
+        {'a': 2}
+
+        'F' key is now lowercase.
+
+        >>> cd['f']
+        6
+
+        '_' is now '-'
+
+        >>> cd['e-5']
+        5
+
+        Keys that have been accessed.
+
+        >>> cd.key_access_stack.keys()
+        dict_keys(['b', 'f', 'e-5'])
```

Comment:
What does this look like if we access `cd['b']['a']`?

Reply:
In this case, you'd get an AttributeError because `a` is just a normal dict. So, if we have a nested dict and we use it to initialize a new `ConfigDict` instance it *will not* automatically cast the nested dicts and assign them parents. This is what `gather_configtree()` is explicitly for.

So, to make this concrete, if we did:

```python
d = {'a': 1, 'b': {'a': 2}, 'c': 3, 'd-4': 4, 'e_5': 5, 'F': 6}
cd = gather_configtree(d)
cd['b']
cd['f']
cd['e-5']
cd['b']['a']
```

Running `cd.key_access_stack.keys()` returns `dict_keys(['b', 'f', 'e-5'])` because it only keeps track of the local keys that were accessed.

Running `cd['b'].key_access_stack.keys()` returns `dict_keys(['a'])`. 

And running `cd['b'].accessed_all_keys()` returns `False`. All the keys in `cd['b']` might have been accessed but not all of the keys in its parent were used. This method climbs up its family tree.